### PR TITLE
DM-25199: Distribute lsst.alert.packet on PyPI

### PR DESCRIPTION
**Pull request by spenczar**
#### Ticket:
[DM-25199](https://jira.lsstcorp.org/browse/DM-25199)

In [DM-25102](https://jira.lsstcorp.org/browse/DM-25102), @ebellm pointed out that we should merge some of the streamsim code (that is, code in github.com/lsst-dm/alert-stream-simulator). That only works if we can still distribute `alert-stream-simulator`, which in turn only works if we can redistribute alert_packet after merging code from there into here, so here we are.

The commit log is pretty verbose on why each change needs to be made, like moving the schema directory and stopping the use of symlinks. The short version is that we need to obey `setuptools`' whims. A lot of the pain is that this package includes non-Python files that need to be distributed.

NB: Nothing has been pushed to PyPI so far. I don't have a clear plan on how to manage PyPI creds, and would like feedback.

### CODE REVIEW

File: test/test_setup.py.sh
Comment (spenczar) [inline test/test_setup.py.sh]: I don't really know a good way to run this, but here ya go, a little bash script test of buildability. This was really important for getting all the `pkg_resources` stuff right.

File: README.rst
Comment (timj) [inline README.rst]: packet or packets? I think packet.

File: bin/validateAvroRoundTrip.py
Comment (timj) [inline bin/validateAvroRoundTrip.py:78]: General comment but there shouldn't be any files in a bin/ directory.  All this code should be part of the module and for pypi should be pulled in as an entry point (defined in setup.cfg)  -- see astro_metadata_translator.  (I was initially confused by the lack of bin.src but I see you aren't using scons at all).

File: python/lsst/alert/packet/schema.py
Comment (timj) [inline python/lsst/alert/packet/schema.py:50]: This returns `bytes`??? That's a very poor name in that case. Wonder what they were thinking.

File: python/lsst/alert/packet/schema.py
Comment (timj) [inline python/lsst/alert/packet/schema.py]: This blank line should not be added.

File: setup.py
Comment (timj) [inline setup.py]: Please move most of this to setup.cfg so that other tooling can find it.

File: setup.py
Comment (timj) [inline setup.py]: It seems like this should be doing `os.path.join(dir, d)` and not assuming `/`

File: setup.py
Comment (timj) [inline setup.py]: I will reiterate my entry points comment here.

File: setup.py
Comment (timj) [inline setup.py]: Out of interest, what is making this False?

**Review by timj — COMMENTED** [review]
This looks okay but:

1. Those scripts in bin should move to entry points.
2. Since you don't depend on anything in the stack please include a (GitHub action I guess) that does the python setup.py install and runs the tests (I don't see any code that automatically runs the tests).

File: python/lsst/alert/packet/schema.py
Comment (jdswinbank) [inline python/lsst/alert/packet/schema.py]: Take that last `.`...

File: python/lsst/alert/packet/schema.py
Comment (jdswinbank) [inline python/lsst/alert/packet/schema.py]: ...and move it here!

File: python/lsst/alert/packet/schema.py
Comment (jdswinbank) [inline python/lsst/alert/packet/schema.py]: If I were @timj , I'd probably mutter something about `os.path.join` here...

**Comment by spenczar** [comment]
Thanks for the super-fast turnaround, @timj.

> Those scripts in bin should move to entry points.

:+1: Sounds right.

> Since you don't depend on anything in the stack please include a (GitHub action I guess) that does the python setup.py install and runs the tests (I don't see any code that automatically runs the tests).

Funny you should mention that, since the tests don't pass on master (https://jira.lsstcorp.org/browse/DM-25208). I don't think anyone noticed since, as you say, they weren't running automatically.

File: setup.py
Comment (spenczar) [reply setup.py]: `setuptools` actually wants _only_ slash-delimited paths here: 
> Also notice that if you use paths, you must use a forward slash (/) as the path separator, even if you are on Windows. Setuptools automatically converts slashes to appropriate platform-specific separators at build time.

(from deep in the bowels of https://setuptools.readthedocs.io/en/latest/setuptools.html#including-data-files)

File: setup.py
Comment (spenczar) [reply setup.py]: Honestly, just reflexive habit. I didn't think about it one bit, I just have been trained over time to assume things aren't zip-safe unless proven otherwise.

Using pkg_resources maybe makes this fine, I'm not sure.

File: schema
Comment (jdswinbank) [inline schema:1]: It feels weird to me to have the schema files themselves buried in the `python` directory. Not dealbreaker-weird, but is it really necessary?

File: setup.py
Comment (timj) [reply setup.py]: Some things are too clever. Of course using `posixpath.join` will ensure that 😉 

Also I tend to prefer f-strings (`f"{dir}/{d}"`)...

File: setup.py
Comment (jdswinbank) [inline setup.py]: Should probably give Vera Rubin her “C.”.

File: setup.py
Comment (timj) [reply setup.py]: Yes. I think pkg_resources means that zip access is transparent. That's one of the great things about it.

**Review by jdswinbank — APPROVED** [review]
Minor quibbles only. I agree with pretty much everything @timj said.

File: python/lsst/alert/packet/schema.py
Comment (spenczar) [reply python/lsst/alert/packet/schema.py:50]: Yep, I immediately stubbed my toe on that too. Maintainers just kind of shrug at it and say "sorry we can't fix it": https://mail.python.org/pipermail/distutils-sig/2011-March/017495.html

Comment (timj) [reply python/lsst/alert/packet/schema.py:50]: The hilarity of someone in 2011 saying that it can't be changed in python3 because there are too many users...

File: schema
Comment (spenczar) [reply schema:1]: See https://github.com/lsst/alert_packet/pull/15/commits/4ae3eebc0167d45073b4abc4cdfb5c2af9a7b2f5 - the docs on "how to add data files that aren't part of a python package" literally just respond with "Instead, the PyPA recommends that any data files you wish to be  accessible at run time be included in the package." I think we're stuck with this, other than a few wild ideas like compiling the schemas directly into Python code (which still strikes me as a good idea!).

File: python/lsst/alert/packet/schema.py
Comment (spenczar) [reply python/lsst/alert/packet/schema.py]: This is for the same reason as discussed in https://github.com/lsst/alert_packet/pull/15#discussion_r434238043. Resource paths _look_ like OS paths, but they aren't actually the same thing. They always use slashes, so os.path.join would be wrong here (see https://setuptools.readthedocs.io/en/latest/pkg_resources.html#basic-resource-access).

Yeah, it's weird.

File: schema
Comment (timj) [reply schema:1]: Yes, you have to put them in the package if you are using pkg_resources. This is the same reason why I had to relocate the astro_metadata_translator corrections YAML files into the python directory.

Comment (jdswinbank) [reply schema:1]: Well, that's sad!

This is fine for the short term. Longer term, we should think about how we actually want to distribute these schema files — there should be one canonical place for them, and it's not _obvious_ that should be coupled to Python software releases. But let's not block on that now.

File: setup.py
Comment (spenczar) [reply setup.py]: I'm curious: why setup.cfg instead of pyproject.toml, https://www.python.org/dev/peps/pep-0518/?

Comment (timj) [reply setup.py]: Probably because I'm an old fuddy duddy who doesn't keep up with all the hip kids. My main point is to move as much of the static configuration code out of python and into configuration files.

File: setup.cfg
Comment (timj) [inline setup.cfg:34]: You might want to hold out on RFC-698 before uploading this to PyPI (although I see the planned end for that is quite long from now) so that you can drop the .py from the executable name.

File: README.rst
Comment (timj) [inline README.rst]: Are we really telling people to add that directory to their PATH? Is this because you don't want people to install it?

Comment (spenczar) [reply README.rst]: I inherited this repo, and don't really understand this section myself; I'm erring on the side of changing one thing at a time. I agree with you, I think. I suspect that this was in the README just because it was expedient, but I don't really know.

Comment (jdswinbank) [reply README.rst]: I think you should assume that all pre-existing installation instructions were based on expediency: there was no meaningful installation system until you added one. :-) I'd suggest removing anything that isn't obviously correct.

File: .github/workflows/build.yml
Comment (timj) [inline .github/workflows/build.yml]: daf_butler has a lint.yaml workflow that you can copy directly. Having distinct workflows for distinct operations seems to be a good way of doing things since they all run in parallel.

File: README.rst
Comment (spenczar) [reply README.rst]: Sorry [Chesterton](https://en.wikipedia.org/wiki/Wikipedia:Chesterton%27s_fence), we're bulldozing all the fences today!

(sounds good to me :))

File: .github/workflows/build.yml
Comment (jdswinbank) [reply .github/workflows/build.yml]: A neat thing is that jobs within a workflow will run in parallel too, unless there are dependencies between them. Github Actions is pretty great.

However... I still agree with Tim here, as it'd be super nice to be able to drop the same `lint.yaml` file into all of our packages, rather than maintaining special-snowflake workflows for each one.

Comment (spenczar) [reply .github/workflows/build.yml]: That's good with me, but I'd like to do that separately. DM-25199 blocks [DM-25104](https://jira.lsstcorp.org/browse/DM-25104), which blocks [DM-24645](https://jira.lsstcorp.org/browse/DM-24645), which blocks [DM-25226](https://jira.lsstcorp.org/browse/DM-25226), which has a real due date of 10 days. That's a deep chain of blockers! Adding linting is important but we should unblock the other work, and then add linting in parallel.

File: setup.cfg
Comment (spenczar) [reply setup.cfg:34]: I think I'm fine with things ending with `.py`, at least for now. It's a little yucky, but not enough to make me want to wait.

**Comment by spenczar** [comment]
@timj, I think I've addressed all your comments. Do you have anything else you consider blocking?

File: .github/workflows/build.yml
Comment (timj) [inline .github/workflows/build.yml]: Can you add a comment here explaining why pip is not new enough?

**Comment by timj** [comment]
Looks fine to me.

File: .github/workflows/build.yml
Comment (spenczar) [reply .github/workflows/build.yml]: Thanks for prompting this. This was a red herring when I was debugging the GH action; the real issue I was hunting was resolved in https://github.com/lsst/alert_packet/pull/15/commits/801fa738bc3991ce3046f99aea443c726babe90d - I didn't specify package_data correctly. I'll remove this upgrade line.

**Pull request merged**
